### PR TITLE
cds-1705 Enhance Metadata collection and dynamic allocation for App/Subsystem Name

### PR DESCRIPTION
**Pull request by daidokoro**
This PR enhances the metadata collection capabilities of the shipper. Metadata is now collected from all available integrations internally and can be added using the AddMetadata parameter of the integration. 

Dynamic Value allocation syntax has been updated and enhanced allowing for the use of regular expressions to parse available metadata values while setting subsystem and application name values.

```
{{ metadata | r'regex' }}
```

- **added metadata context type**
- **implemented metadataContext type to handle metadata cross functinos and threads**
- **cargo**
- **updated JSONMessage type with new attribute names, added support functions to struct**
- **added additoinal metadata for sqs, ecr and kinesis events**
- **udpatd test**
- **modified metadata key names and updated MetadataContext struct**
- **fixed dynamic metadata replacement syntax**
- **add tests for metadataContext evaluation**
- **removed headers from LogExporterConfig as the option has been removed in upstream sdk**
- **added MetadataContext method to handle dynamic metadata allocation and removed old tests**
- **cargo update**
- **changelog**
- **updated README**
- **adjusted metadata name for s3.key to s3.object.key**
- **template version bump**
- **version bump**

## Important

This PR contains breaking changes

# Description

 Fixes #

# How Has This Been Tested?

# Checklist:
- [ ] I have updated the versions in the SemanticVersion in template.yaml
- [ ] I have updated the CHANGELOG.md
- [ ] I have created necessary PR to Terraform Module Repository (https://github.com/coralogix/terraform-coralogix-aws) if needed
- [ ] This change does not affect any particular component (e.g. it's CI or docs change)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -222,6 +221,74 @@ Use the following options if you need to configure a private link with Coralogix
...
+Some metadata keys will overlap as some integrations share the same metadata. For example, both Kafka and MSK have the same metadata key `kafka.topic` or both Kinesis and Cloudwatch metadata will be added in cases where a Cloudwatch log stream is being ingested from a Kinesis stream.
+
+##### Dynamic Subsystem or Application Name
+
+As of `v1.1.0`,you can use dynamic values for the Application and Subsystem Name parameters based on the internal metadata defined above.
+
+To do accomplish this, you can use the following syntax:
+
+```
+ApplicationName/SubsystemName: {{ metadata.key | r'regex' }}
```
Comment (RafalSumislawski) [inline README.md:261]: IMO this example with a `/` is confusing. It would be cleared to have separate examples for `ApplicationName` and `SubsystemName`

File: README.md [hunk 2]
```diff
@@ -222,6 +221,74 @@ Use the following options if you need to configure a private link with Coralogix
...
+
+This would result in a SubsystemName value of `elb.log` as this is the part of the regex that is captured by the group `(.*)`.
+
+**Important**:
+
+- The regex must be a valid regex pattern.
+- The regex must define a capture group for part of the string you want to use as the value
+- The metadata key must exist in the list defined above and be a part of the integration type that is deployed.
+
+This would extract the last part of the bucket name after the last `/`.
```
Comment (RafalSumislawski) [inline README.md:290]: This line is out of place. What the example does is already explained in line 282

File: src/logs/mod.rs [hunk 3]
```diff
@@ -187,18 +193,38 @@ pub async fn handler(
                         result?;
                     } else {
                         debug!("SQS TEXT EVENT Detected");
-                        process::sqs_logs(message.clone(), coralogix_exporter.clone(), config)
-                            .await?;
+                        process::sqs_logs(
+                            &mctx,
+                            message.clone(),
```
Comment (RafalSumislawski) [inline src/logs/mod.rs:198]: Not really part of this PR but I think we wouldn't need to clone here if we didn't use references in lines 109 and 105

File: src/logs/process.rs [hunk 4]
```diff
@@ -11,34 +11,124 @@ use fancy_regex::Regex;
...
+    /// values are expected to be in the format of<br>
+    /// `{{key}}` or `{{key|regex}}`, the regex must include a capture group which will be
+    /// used to extract the value. The key is the key in the metadata context map.
+    pub fn evaluate(&self, value: String) -> Result<String, String> {
+        if !value.starts_with("{{") && !value.ends_with("}}") {
+            return Ok(value);
+        };
+
+        let (reg, key) = if value.contains("|") {
+            let r = Regex::new(r#"\{\{\s?(?<key>[a-z\.0-9]+)\s?\|?\s?r'(?<regex>.*)'\s?\}\}"#)
```
Comment (RafalSumislawski) [inline src/logs/process.rs:65]: Recreating these regexes on each invocation of `evaluate` may be costly. I would recommend using something like `lazy_static` or `LazyLock` to create them once in a static variable.

Comment (RafalSumislawski) [inline src/logs/process.rs:65]: IMO these `\s?` could be `\s*`. Two spaces don't look much different from one space and if we accept the expression with 0 or 1 space I see no reason to reject it when it has 2 spaces.

Comment (RafalSumislawski) [inline src/logs/process.rs:65]: The `[a-z\.0-9]+` will not accept `_` and we have a `kinesis.event.source_arn` in the documentation

File: src/logs/process.rs [hunk 5]
```diff
@@ -317,9 +351,10 @@ pub async fn cloudwatch_logs(
 
     let logs = match config.integration_type {
         IntegrationType::CloudWatch => {
-            metadata_instance.stream_name = cloudwatch_event_log.data.log_stream.clone();
-            metadata_instance.log_group = cloudwatch_event_log.data.log_group.clone();
+            // metadata_instance.stream_name = cloudwatch_event_log.data.log_stream.clone();
```
Comment (RafalSumislawski) [inline src/logs/process.rs:354]: This code could probably be deleted

File: Cargo.toml [hunk 6]
```diff
@@ -38,6 +38,7 @@ async-recursion = "1.1.0"
 md5 = "0.7.0"
 anyhow = "1.0.86"
 aws-sdk-sts = "1.49.0"
+handlebars = "6.2.0"
```
Comment (RafalSumislawski) [inline Cargo.toml:41]: Is this actually used?

**Review by RafalSumislawski — APPROVED** [review]

**Comment by daidokoro** [comment]
Thanks for the review @RafalSumislawski, I've fixed all the items you've pointed out. Let me know if you spot anything else.
